heave_ho.asl

```diff
@@ -1,147 +1,176 @@
-    //IntPtr levelManager = memory.ReadPointer(currentCharacter+0x100);
-    //IntPtr levelRules = memory.ReadPointer(levelManager+0x20);
-    //IntPtr victoryTrigger = memory.ReadPointer(levelManager+0x30);
-    //current.IsVictory = memory.ReadValue<bool>(victoryTrigger+0x88);
-    //print(player.ToString());
-    IntPtr levelTimes = new IntPtr(vars.LevelTimesPtr.Deref<long>(game));
-    //print(levelTimes.ToString("X16"));
-    current.LevelTimesLength = memory.ReadValue<int>(levelTimes+0x18);
-    vars.LevelTimes = memory.ReadPointer(levelTimes+0x10);
-}
-
-gameTime
-{
-    if (current.LevelTimesLength < old.LevelTimesLength)
-    {
-        vars.TotalRunTime += vars.CurrentRunTime;
-    }
-    
-    float runTime = 0.0f;
-    for (int i = 0; i < current.LevelTimesLength; i++)
-    {
-        runTime += memory.ReadValue<float>( ((IntPtr)vars.LevelTimes) + 0x20 + (4 * i));
-    }
-    vars.CurrentRunTime = runTime;
-    if (vars.JustStarted)
-    {
-        vars.JustStarted = false;
-        vars.offset = vars.CurrentRunTime;
-    }
-    //if(current.IsVictory)
-    //{
-        return TimeSpan.FromSeconds((double)(vars.TotalRunTime + vars.CurrentRunTime) - vars.offset);
-    //}
-}
-
-start
-{
-    //return current.LevelName == "Tuto1" && old.LevelName != "Tuto1";
-}
-
-split
-{
-    return current.LevelTimesLength > old.LevelTimesLength;
-}
-
-isLoading
-{
-    return true;
-}
\ No newline at end of file
+state("HeaveHo") {}
+
+startup
+{
+    vars.PAGE_EXECUTE_ANY = MemPageProtect.PAGE_EXECUTE | MemPageProtect.PAGE_EXECUTE_READ | MemPageProtect.PAGE_EXECUTE_READWRITE | MemPageProtect.PAGE_EXECUTE_WRITECOPY;
+
+    vars.JustStarted = false;
+    vars.offset = 0.0;
+    vars.TotalRunTime = 0.0f;
+    vars.CurrentRunTime = 0.0f;
+
+    vars.OnStart = (EventHandler) (
+        (object sender, EventArgs e) => {
+            vars.JustStarted = true;
+            vars.offset = 0.0;
+            vars.TotalRunTime = 0.0;
+        }
+    );
+
+    timer.OnStart += vars.OnStart;
+}
+
+shutdown
+{
+    timer.OnStart -= vars.OnStart;
+}
+
+init
+{
+    // Version checking using hash
+    vars.gamePath = modules.First().FileName + "\\..\\HeaveHo_Data\\Managed\\Assembly-CSharp.dll";
+    byte[] dllBytes = File.ReadAllBytes(vars.gamePath);
+    System.Security.Cryptography.SHA256 hasher = System.Security.Cryptography.SHA256.Create();
+    byte[] hashed = hasher.ComputeHash(dllBytes);
+    string s = "";
+    foreach (byte b in hashed) s += b;
+    if (s == "2336514740711751125296795513677151228992114533192051056830181341724411522967") version = "1.0";
+    else if (s == "2308696111144216968593416917814132145111418649485985105170732501369717521317792") version = "1.1";
+    else version = "NA";
+    print("[ASL] Hash: " + s);
+    print("[ASL] Version: " + version);
+    
+    // Error out if unknown version
+    if (version == "NA") {
+        print("[ASL] Unknown game hash / version");
+        print("[ASL] " + s);
+        return;
+    }
+
```

Comment:
unknown version check returns out of init. Will stop getting stuck on while loop.